algorithm/cgal/occt_util.h

```diff
@@ -0,0 +1,188 @@
+#include "BRepPrimAPI_MakeBox.hxx"
+#include "BRepPrimAPI_MakeSphere.hxx"
+#include "BRepTools.hxx"
+#include "GC_MakeSegment.hxx"
+#include "RWMesh_FaceIterator.hxx"
+#include "StdFail_NotDone.hxx"
+#include "TopoDS.hxx"
+#include "TopoDS_Builder.hxx"
+#include "gp_Trsf.hxx"
+
+TopoDS_Solid sewOcctSolidFromSurfaceMesh(const Surface_mesh& mesh) {
+  BRepBuilderAPI_Sewing builder;
+  for (const auto& facet : mesh.faces()) {
+    const auto& start = mesh.halfedge(facet);
+    if (mesh.is_removed(start)) {
+      continue;
+    }
+    Halfedge_index edge = start;
+    BRepLib_MakeWire wire_maker;
+    do {
+      Point cgal_s = mesh.point(mesh.source(edge));
+      Point cgal_t = mesh.point(mesh.target(edge));
+      gp_Pnt occt_s(CGAL::to_double(cgal_s.x().exact()), CGAL::to_double(cgal_s.y().exact()), CGAL::to_double(cgal_s.z().exact()));
+      gp_Pnt occt_t(CGAL::to_double(cgal_t.x().exact()), CGAL::to_double(cgal_t.y().exact()), CGAL::to_double(cgal_t.z().exact()));
+      Handle(Geom_TrimmedCurve) occt_segment = GC_MakeSegment(occt_s, occt_t);
+      wire_maker.Add(BRepBuilderAPI_MakeEdge(occt_segment));
+      edge = mesh.next(edge);
+    } while (edge != start);
+    builder.Add(BRepBuilderAPI_MakeFace(wire_maker.Wire()));
+  }
+  builder.Perform();
+  BRepBuilderAPI_MakeSolid solid_maker;
+  solid_maker.Add(TopoDS::Shell(builder.SewedShape()));
+  return solid_maker.Solid();
+}
+
+TopoDS_Solid buildOcctSolidFromSurfaceMesh(const Surface_mesh& mesh) {
+  BRep_Builder builder;
+  TopoDS_Shell shell;
+  builder.MakeShell(shell);
+  for (const auto& facet : mesh.faces()) {
+    const auto& start = mesh.halfedge(facet);
+    if (mesh.is_removed(start)) {
+      continue;
+    }
+    Halfedge_index edge = start;
+    BRepLib_MakeWire wire_maker;
+    do {
+      Point cgal_s = mesh.point(mesh.source(edge));
+      Point cgal_t = mesh.point(mesh.target(edge));
+      gp_Pnt occt_s(CGAL::to_double(cgal_s.x().exact()), CGAL::to_double(cgal_s.y().exact()), CGAL::to_double(cgal_s.z().exact()));
+      gp_Pnt occt_t(CGAL::to_double(cgal_t.x().exact()), CGAL::to_double(cgal_t.y().exact()), CGAL::to_double(cgal_t.z().exact()));
+      Handle(Geom_TrimmedCurve) occt_segment = GC_MakeSegment(occt_s, occt_t);
+      wire_maker.Add(BRepBuilderAPI_MakeEdge(occt_segment));
+      edge = mesh.next(edge);
+    } while (edge != start);
+    builder.Add(shell, BRepBuilderAPI_MakeFace(wire_maker.Wire()));
+  }
+  BRepBuilderAPI_MakeSolid solid_maker;
+  solid_maker.Add(shell);
+  return solid_maker.Solid();
+}
+
+void buildSurfaceMeshFromOcctShape(const TopoDS_Shape& shape, double deflection_tolerance, Surface_mesh& mesh) {
+  Vertex_map vertices;
+  for(TopExp_Explorer it(shape, TopAbs_SOLID) ; it.More() ; it.Next()){
+    TopoDS_Solid solid = TopoDS::Solid(it.Current());
+    BRepMesh_IncrementalMesh(solid, deflection_tolerance);
+    // Each face is triangulated separately.
+    // We could try to consolidate vertices, but let's see if we get t-junctions first.
```

Comment:
How did this play out?

Reply:
Ah, that needs updating -- we don't see t junctions because the triangulation is entirely controlled by geometric shape, unlike in jscad.